[PGO] Work around memory corruption in instrumentation runtime

emw-apple opened the pull request:
#### 987f734c374ec6d7091385529cf241015177934d
<pre>
[PGO] Work around memory corruption in instrumentation runtime
<a href="https://bugs.webkit.org/show_bug.cgi?id=292899">https://bugs.webkit.org/show_bug.cgi?id=292899</a>
<a href="https://rdar.apple.com/151191066">rdar://151191066</a>

Reviewed by Alexey Proskuryakov.

When building with instrumentation, move the __llvm_prf_cnts section to
its own segment. This section is the table of function counters
generated for a binary. At runtime, the compiler-rt library mmaps this
section to the .profdata file for that framework and process.

However, it rounds up to the nearest page boundary, which may be past
the end of the section. When the counts table is not itself
page-sized, the call to mmap might write over another section in __DATA,
such as ObjC runtime data (<a href="https://rdar.apple.com/151083138">rdar://151083138</a>).

We encounter this as `NSProtocolFromString(@&quot;WKObject&quot;)` returning
garbage data in some instrumented builds of WebKit. But it could happen
to anything in __DATA. By moving __llvm_prf_cnts to its own segment, it
is guaranteed to be the only section in its page and can mmap freely.

* Configurations/CommonBase.xcconfig:

Canonical link: <a href="https://commits.webkit.org/294823@main">https://commits.webkit.org/294823@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/202a87b7b31920a4401bb8f0a24078c47bc64fef

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103231 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22907 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13227 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108401 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53874 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23243 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31407 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78460 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35391 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106237 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18020 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93133 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58793 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/17861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53229 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/87668 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11238 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110777 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30369 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/22385 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87458 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30734 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89332 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87091 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/22156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31949 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9664 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24692 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16745 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30298 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/30104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33431 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/31666 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov reviewed (approved):
r=me with due diligence testing.

emw-apple committed:
[PGO] Work around memory corruption in instrumentation runtime
https://bugs.webkit.org/show_bug.cgi?id=292899
rdar://151191066

Reviewed by Alexey Proskuryakov.

When building with instrumentation, move the __llvm_prf_cnts section to
its own segment. This section is the table of function counters
generated for a binary. At runtime, the compiler-rt library mmaps this
section to the .profdata file for that framework and process.

However, it rounds up to the nearest page boundary, which may be past
the end of the section. When the counts table is not itself
page-sized, the call to mmap might write over another section in __DATA,
such as ObjC runtime data (rdar://151083138).

We encounter this as `NSProtocolFromString(@"WKObject")` returning
garbage data in some instrumented builds of WebKit. But it could happen
to anything in __DATA. By moving __llvm_prf_cnts to its own segment, it
is guaranteed to be the only section in its page and can mmap freely.

* Configurations/CommonBase.xcconfig:

Canonical link: https://commits.webkit.org/294823@main
